### DISCUSSION TITLE
How to write a html code of table or text to a define rectangular area?

### DISCUSSION THREAD

**Discussion by nissansz** [pymupdf/PyMuPDF#4059] [Q&A]
How to write a html code of table or text to a define rectangular area?

**Comment by JorjMcKie** [comment]
This is no issue but a question.
Use [insert_htmlbox](https://pymupdf.readthedocs.io/en/latest/page.html#Page.insert_htmlbox).

**Reply by nissansz** [reply]
If I want to use a white background before writing html, use below method?

draw_rect(rect, color=(0,), fill=None, width=1, dashes=None, lineCap=0, lineJoin=0, overlay=True, morph=None, stroke_opacity=1, fill_opacity=1, radius=None, oc=0)
PDF only: Draw a rectangle. See [Shape.draw_rect()](https://pymupdf.readthedocs.io/en/latest/shape.html#Shape.draw_rect).

**Reply by JorjMcKie** [reply]
Use `page.draw_rect(rect, fill=(1,1,1),color=None)`.
This fills an area with white and draws no border.

**Reply by nissansz** [reply]
how to support morph=(pos, mat) for insert_htmlbox?

**Reply by nissansz** [reply]
How to set the rect to be half-transparent?

**Reply by nissansz** [reply]
page.draw_rect(rect, color=None, fill=fitz.pdfcolor["lightblue"], opacity=0.5, overlay=False)
TypeError: draw_rect() got an unexpected keyword argument 'opacity'

**Reply by JorjMcKie** [reply]
Looking at the documentation is always a good idea 😎:
![image](https://github.com/user-attachments/assets/eb81d668-d006-4762-9961-3be5c2bf96ee)

**Reply by nissansz** [reply]
what is the official link for documentation?

**Reply by jamie-lemon** [reply]
https://pymupdf.readthedocs.io/en/latest/page.html#Page.draw_rect

**Reply by nissansz** [reply]
Thank you. Any textline detection model in pymupdf?

**Reply by nissansz** [reply]
Is there any method to paste an cv2 image to rect?

Below method seems supporting filename only.

insert_image(rect, *, alpha=-1, filename=None, height=0, keep_proportion=True, mask=None, oc=0, overlay=True, pixmap=None, rotate=0, stream=None, width=0, xref=0)

**Reply by JorjMcKie** [reply]
You have 4 alternatives:
![image](https://github.com/user-attachments/assets/c3c16282-6436-433c-a23e-5f170f1c8288)

* Use an xref of some image **_already_** inserted
* Use a filename
* Use a pixmap
* Use a `bytes`-like memory area (stream parameter)

**Reply by nissansz** [reply]
How to convert cv2 img to pixmap ?

set pixmap = converted img from cv2?

**Reply by JorjMcKie** [reply]
I almost never work with cv2, but it should have a conversion or export method cv2-image to `bytes`. Probably something like this:
```python

# Encode the image as a JPEG
_, buffer = cv2.imencode('.jpg', image)

# Convert the buffer to bytes
image_bytes = buffer.tobytes()

page.insert_image(..., stream=image_bytes, ...)
```
Using this alternative is more practical, ...

**Reply by nissansz** [reply]
It works, thanks.

**Comment by JorjMcKie** [comment]
Not implemented

Gesendet von Outlook für Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by nissansz** [reply]
When will it be implemented?

**Reply by JorjMcKie** [reply]
No plans to do this at all currently.

**Comment by nissansz** [comment]
span [color](background-color: cannot be shown after inserted,
<span style="background-color: aqua;"

but below code can be displayed.
<span style="color: green;font-weight:bold;"

**Reply by JorjMcKie** [reply]
Background color has no effect on inline-level elements like span. Use block-level elements like `<div>` or `<p>`.

**Comment by nissansz** [comment]
If original pdf is flow text, when I wrote html overlay on original area, 
when I select text from pdf, only the original flow text, how to select my written text?

**Reply by JorjMcKie** [reply]
Don't understand: what is "flow text"?
Otherwise: extracting text from a given rectangle extracts **_always all_** text in that area.

**Comment by nissansz** [comment]
flow text is editable source text or ocr layer by other software.
I manually copy textline text

![image](https://github.com/user-attachments/assets/94ac2506-32f4-4b67-8a0a-7456b47403f5)
ocr layer copy result: 230官峨o…官1をワークメモリとして聞いる構成とし

Sometimes, if my newly added text layer above original flow text overlapps with original flow text, when I copy the textline,
only the old text line can be copied. I may want to copy the newly added text shown.


[p5-1_300_ocr.pdf](https://github.com/user-attachments/files/17909903/p5-1_300_ocr.pdf)

**Comment by nissansz** [comment]
If original paper orientation is horizontal / Landscape, how to set parameters?
    spage.remove_rotation()  # remove rotation - keeping visual appearance
    page = out.new_page()  # make A4 output page

**Reply by nissansz** [reply]
I can get corret orientation with below setting.
'''
    spage.remove_rotation()  # remove rotation - keeping visual appearance
    # page = out.new_page()  # make A4 output page
    page = out.new_page( width=pix_original.width, height=pix_original.height)
'''

**Comment by nissansz** [comment]
Any method to stretch the inserted html box like a normal textwriter?

**Reply by JorjMcKie** [reply]
No, there is no way to do that.

**Reply by nissansz** [reply]
Any chance to add morph change function for html box?

**Reply by JorjMcKie** [reply]
This is not planned.

**Reply by nissansz** [reply]
Are these same? Any difference?
  doc = fitz.open(path)
  doc = pymupdf.open(path)

**Reply by JorjMcKie** [reply]
I hate to say "NO" all the time, but you **_can_** influence our internal priority settings!
We plan our feature enhancements primarily according to economic principles and market trends.
Which, among other implications, also means that enhancement requests issued by our commercial clients are served with top priority.

So, depending on what you do or for whom you work, best contact our [sales and marketing department](https://artifex.com/contact/) to begin a discussion about acquiring a commercial license.
There is no fixed price list: every commercial license is negotiated individually, taking into account the type business type, customers, usage profile (e.g. number of servers) and other criteria. Also included are special startup company considerations.

**Reply by nissansz** [reply]
ok
